scheds/rust/scx_layered/src/bpf/main.bpf.c

```diff
@@ -560,30 +560,42 @@ void BPF_STRUCT_OPS(layered_dispatch, s32 cpu, struct task_struct *prev)
 
 	/* consume preempting layers first */
 	bpf_for(idx, 0, nr_layers)
-		if (layers[idx].preempt && scx_bpf_consume(idx))
```

Comment:
I'm not sure if `layered_dispatch` is the right place to put this, so please let me know if there is a more suitable place. Probably could have cleaned up some of the redundant checks as well.

Reply:
I think we'll want to put this in `layered_running()`. That callback is invoked just before a task starts running on a CPU, so we can look at what layer it's in and apply the frequency tuning accordingly.

As an aside, `layered_dispatch()` isn't the only place where we can choose where a task will run. `layered_select_cpu()`, which is called on the task wakeup (and task fork) path, decide which CPU a task should be migrated to before being enqueued. We can "directly dispatch" to local DSQs from there, which is conceptually the same thing as calling `scx_bpf_consume()`. See `scx_bpf_dispatch(p, SCX_DSQ_LOCAL, ...)`. In the former case, we're putting a task directly on the local DSQ where the core kernel will pull the task from to run, and in the latter case, we're taking a task from another DSQ and putting it on the local DSQ. The end result is the same -- we're specifying a task that should run next on that CPU. By instead just looking up a task's layer from `layered_running()`, we're guaranteed to hit all the cases.

If we want to speed up the context switch path, we could also reset the frequency in `layered_stopping()`, but that could have the downside of accidentally throttling a CPU that's running a high-pri layer task.